src/main/java/com/amazonaws/encryptionsdk/keyrings/EncryptionMaterials.java

```diff
@@ -0,0 +1,237 @@
+import com.amazonaws.encryptionsdk.EncryptedDataKey;
+
+import javax.crypto.SecretKey;
+import java.security.PrivateKey;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.apache.commons.lang3.Validate.isTrue;
+import static org.apache.commons.lang3.Validate.notNull;
+
+/**
+ * Contains the cryptographic materials needed for an encryption operation with Keyrings.
+ */
+public final class EncryptionMaterials {
+    private final CryptoAlgorithm algorithm;
+    private final Map<String, String> encryptionContext;
+    private final List<EncryptedDataKey> encryptedDataKeys;
+    private SecretKey plaintextDataKey;
+    private final PrivateKey signingKey;
+    private final KeyringTrace keyringTrace;
+
+    private EncryptionMaterials(Builder b) {
+        notNull(b.algorithm, "algorithm is required");
+        notNull(b.keyringTrace, "keyringTrace is required");
+        validatePlaintextDataKey(b.algorithm, b.plaintextDataKey);
+        validateSigningKey(b.algorithm, b.signingKey);
+        this.algorithm = b.algorithm;
+        this.encryptionContext = b.encryptionContext;
+        this.encryptedDataKeys = b.encryptedDataKeys;
+        this.plaintextDataKey = b.plaintextDataKey;
+        this.signingKey = b.signingKey;
+        this.keyringTrace = b.keyringTrace;
+    }
+
+    public Builder toBuilder() {
+        return new Builder(this);
+    }
+
+    public static Builder newBuilder(CryptoAlgorithm algorithm) {
+        return new Builder(algorithm);
+    }
+
+    /**
+     * The algorithm suite to be used for encryption.
+     */
+    public CryptoAlgorithm getAlgorithm() {
+        return algorithm;
+    }
+
+    /**
+     * The encryption context associated with this encryption.
+     */
+    public Map<String, String> getEncryptionContext() {
+        return encryptionContext;
+    }
+
+    /**
+     * An unmodifiable list of the encrypted data keys that correspond to the plaintext data key.
+     */
+    public List<EncryptedDataKey> getEncryptedDataKeys() {
+        return Collections.unmodifiableList(encryptedDataKeys);
+    }
+
+    /**
+     * Add an encrypted data key to the list of encrypted data keys.
+     *
+     * @param encryptedDataKey  The encrypted data key to add.
+     * @param keyringTraceEntry The keyring trace entry recording this action.
+     */
+    public void addEncryptedDataKey(EncryptedDataKey encryptedDataKey, KeyringTraceEntry keyringTraceEntry) {
+        notNull(encryptedDataKey, "encryptedDataKey is required");
+        notNull(keyringTraceEntry, "keyringTraceEntry is required");
+        encryptedDataKeys.add(encryptedDataKey);
+        keyringTrace.add(keyringTraceEntry);
+    }
+
+    /**
+     * A data key to be used as input for encryption.
+     */
+    public SecretKey getPlaintextDataKey() {
+        return plaintextDataKey;
```

Comment:
This may not be as important for decrypt, but trying to get the plaintext data key before it is set is really bad.

Reply:
Throws an IllegalStateException now